floss/language/go/extract.py

```diff
@@ -0,0 +1,471 @@
+    0000000000492A60  08 00 00 00 00 00 00 00  08 00 00 00 00 00 00 00  ................
+    """
+
+    # Get the blob
+    blob_end_pattern = re.compile(b"\x00{8}.\x00{7}.\x00{7}")
+    blob_end = re.search(blob_end_pattern, section_data[2:])
+    if not blob_end:
+        return
+    blob = section_data[2 : blob_end.start()]
+
+    # Extract strings from the blob
+    blob_pattern = re.compile(b"(\x00|\x01|\x02|\x03)(?P<string_length>.)", re.DOTALL)
+    for m in blob_pattern.finditer(blob):
+        if m.group("string_length") != b"\x00":
+            try:
+                addr = pe.get_offset_from_rva(m.start() + section_va + 4)
+                binary_string = blob[m.end() : m.end() + m.group("string_length")[0]]
+                try:
+                    string = StaticString.from_utf8(binary_string, addr, min_length)
+                    yield string
+                except ValueError:
+                    pass
+            except pefile.PEFormatError:
+                pass
+
+
+def extract_string_blob2(pe: pefile.PE, section_data, section_va, min_length) -> Iterable[StaticString]:
+    # Extract string blob in .rdata section that starts with "go:buildid" or "go.buildid"
+    """
+    67 6F 3A 62 75 69 6C 64  69 64 00 69 6E 74 65 72  go:buildid.inter
+    6E 61 6C 2F 63 70 75 2E  49 6E 69 74 69 61 6C 69  nal/cpu.Initiali
+    7A 65 00 69 6E 74 65 72  6E 61 6C 2F 63 70 75 2E  ze.internal/cpu.
+    70 72 6F 63 65 73 73 4F  70 74 69 6F 6E 73 00 69  processOptions.i
+    6E 74 65 72 6E 61 6C 2F  63 70 75 2E 69 6E 64 65  nternal/cpu.inde
+    78 42 79 74 65 00 69 6E  74 65 72 6E 61 6C 2F 63  xByte.internal/c
+    """
+
+    # Get the blob
+    blob_start_pattern = re.compile(b"go(\.|:)buildid\x00")
+    blob_start = re.search(blob_start_pattern, section_data)
+    if not blob_start:
+        return
+    blob = section_data[blob_start.start() :]
+    blob_end_pattern = re.compile(b"\x00{2}")
+    blob_end = re.search(blob_end_pattern, blob)
+    if not blob_end:
+        return
+    blob = blob[: blob_end.start()]
+
+    # Extract strings from the blob
+    length = 0
+    for binary_string in blob.split(b"\x00"):
+        addr = pe.get_offset_from_rva(blob_start.start() + section_va + length)
+        length += len(binary_string) + 1
+        try:
+            string = StaticString.from_utf8(binary_string, addr, min_length)
+            yield string
+        except ValueError:
+            pass
+
+
+def extract_file_path_strings(pe: pefile.PE, section_data, section_va, min_length) -> Iterable[StaticString]:
+    # 004D14D0  00 1C 00 00 5F 1C 00 00  00 00 00 00 00 00 00 00  ...._...........
+    # 004D14E0  2F 75 73 72 2F 6C 6F 63  61 6C 2F 67 6F 2F 73 72  /usr/local/go/sr
+    # 004D14F0  63 2F 69 6E 74 65 72 6E  61 6C 2F 63 70 75 2F 63  c/internal/cpu/c
+    # 004D1500  70 75 2E 67 6F 00 2F 75  73 72 2F 6C 6F 63 61 6C  pu.go./usr/local
+    # 004D1510  2F 67 6F 2F 73 72 63 2F  69 6E 74 65 72 6E 61 6C  /go/src/internal
```

Comment:
this seems specific to a few samples, can you ensure this is helpful for the general case?

Reply:
doesn't the traditional strings algorithm extract these strings just fine, since they're null terminated?

Reply:
i think this is a really good idea. when we add more code and logic, then there's also more surface area for bugs or mistakes. so i think we should rely on the proven strings.exe algorithm when possible, and only try to parse the things that aren't otherwise recognized (like the big UTF-8 string blob and inline constant strings).

Reply:
Agreed!